src/TaskProviderDune.ml

```diff
@@ -0,0 +1,95 @@
+open Import
+
+let taskDefinition = { Task.type_ = "dune" }
+
+let source = "dune"
+
+let problemMatchers = [| "ocamlc" |]
+
+(* the ocamlc matcher is not able to parse ocaml compiler errors unless they
+   follow the short style. *)
+let env = Js.Dict.fromList [ ("OCAML_ERROR_STYLE", "short") ]
+
+let startsWith ~s ~prefix =
+  try
+    let prefixLen = String.length prefix in
+    let start = String.sub s 0 prefixLen in
+    String.equal start prefix
+  with _ -> false
+
+let removePrefix ~s ~prefix =
+  let prefixLen = String.length prefix in
+  let len = String.length s - prefixLen in
+  String.sub s prefixLen len
+
+let folderRelativePath folders file =
+  Array.fold_left
+    (fun acc (folder : Folder.t) ->
+      match acc with
+      | Some _ -> acc
+      | None -> (
+        match startsWith ~s:file ~prefix:folder.uri.fsPath with
+        | false -> acc
+        | true -> Some (folder, removePrefix ~s:file ~prefix:folder.uri.fsPath)
+        ))
+    None folders
+
+let commandLine () =
```

Comment:
Perhaps we should assume that:

* the dune binary exists 
* there's a dune-project/dune-workspace in the workspace

before providing these tasks.

Reply:
> the dune binary exists

I guess I can, but not sure to see the benefit of it

> there's a dune-project/dune-workspace in the workspace

if I take the example of ahrefs, we have a mono repo with a single dune-project file at the root and multiple projects as sub folders. The rational being to share some config between all projects (like compilation flags). I think there are some people who open one instance of vscode per project. In this case there is no dune-project/dune-workspace in the workspace.

Reply:
> I guess I can, but not sure to see the benefit of it

What's the current behavior if dune is not installed?

> I think there are some people who open one instance of vscode per project. In this case there is no dune-project/dune-workspace in the workspace.

You mean at the workspace root, right? They ought to have to dune-project files in the subdirs. Okay, this isn't essential.

Reply:
> What's the current behavior if dune is not installed?

The build tasks are listed. There would be an error only if one tries to run a task. There are a few task providers in vscode itself. In [gulp](https://github.com/microsoft/vscode/blob/ea0880611f706e0529f89aa25d1989363ab53b11/extensions/gulp/src/main.ts) and [npm](https://github.com/microsoft/vscode/blob/e77c58dd0b6d955c331a3c11321eccae710ecb85/extensions/npm/src/tasks.ts) they don't seem to check for the binary.

To me it seems more natural to have it this way rather than silently doing nothing.

For the record this is how it looks like when one executes `Run Task` in a project with a build tasks configured by hand. The first line is for a task I configured by hand. The ones from an extension are under a directory. Which may or may not by empty.
![image](https://user-images.githubusercontent.com/974142/84427817-19501680-ac58-11ea-9e47-6f502b4c7aee.png)

If I have a build task configured by hand in .vscode/tasks.json, then the tasks from the dune provider are not displayed when executing `Run Build Task`
![image](https://user-images.githubusercontent.com/974142/84427716-ea39a500-ac57-11ea-9943-c57fc1fd093d.png)

It seems that the noise is pretty low.

> You mean at the workspace root, right?

No dune-project/workspace file in the vscode workspace. With the vscode workspace being a project/a subdir if only one project is opened in vscode

Reply:
actually I checked a bit more what happens with npm. With this package.json file. The build tasks appears when executing `Run Build Task`. And I don't have npm installed.

```json
{
    "scripts": {
        "demo-script": "some command",
        "test": "foo",
        "build": "bar"
    }
}
```

![image](https://user-images.githubusercontent.com/974142/84428751-98921a00-ac59-11ea-94db-f5ce19b63ea4.png)